adam/learner/perception_graph_template.py

```diff
@@ -222,3 +225,55 @@ def copy_with_temporal_scopes(
             ),
             template_variable_to_pattern_node=self.template_variable_to_pattern_node,
         )
+
+    def copy_replacing_nodes(
+        self, current_to_new_node: Mapping[NodePredicate, NodePredicate]
+    ) -> "PerceptionGraphTemplate":
+        """
+        Creates a copy of the perception graph template with the pattern updated to align to the new
+        nodes passed in via the mapping.
+        """
+        return PerceptionGraphTemplate(
+            graph_pattern=self.graph_pattern.copy_replacing_nodes(current_to_new_node),
+            template_variable_to_pattern_node=self.template_variable_to_pattern_node,
+        )
+
+    def copy_remove_temporal_scopes(self) -> "PerceptionGraphTemplate":
+        if not self.graph_pattern.dynamic:
+            return self
+
+        return PerceptionGraphTemplate(
+            graph_pattern=self.graph_pattern.copy_remove_temporal_scopes(),
+            template_variable_to_pattern_node=self.template_variable_to_pattern_node,
+        )
+
+    def add_predicate_node(
+        self, predicate_node_to_add: NodePredicate
+    ) -> "PerceptionGraphTemplate":
+
+        root_node = None
+
+        for node in self.graph_pattern:
+            if isinstance(node, type(predicate_node_to_add)) and node.matches_predicate(
+                predicate_node_to_add
+            ):
+                return self  # We are trying to add a duplicate predicate which will do nothing
+
+            if isinstance(node, AnyObjectPredicate):
+                root_node = node
+
+        if root_node is None:
+            raise ValueError(
+                f"Given graph pattern {self.graph_pattern} did not have an object root node."
+            )
+
+        return PerceptionGraphTemplate(
+            graph_pattern=add_predicate_to_pattern_graph(
+                self.graph_pattern,
+                root_node,
+                predicate_node_to_add,
+                temporal_scope=TemporalScope.BEFORE
+                if self.graph_pattern.dynamic
+                else None,
+            )
+        )
```

Comment:
This method is specific enough to the object learner's affordance processing (the specific scope; the assumption that we're dealing with a single-object pattern) that I think it should live somewhere closer to the learner. It doesn't require any protected access to the template, so moving it out should work fine.

I suggest putting this in `learner_utils.py` near the two other object-related functions, `get_objects_from_perception()` and `candidate_object_hypotheses()`. The name should probably change with the move to (for example) `add_predicate_node_to_object_template(...)`. Also, this could use a docstring.